# 01_异常概述

异常：在程序执行过程中，出现的非正常情况，如果不处理最终会导致 JVM 非正常终止。

> 异常不是语法错误或代码逻辑错误。

异常的抛出机制：Java 中把不同的异常用不同的类表示，一旦发生某种异常，就创建该异常类型的对象 ，并且抛出 (throw)。然后程序员可以捕获 (catch) 到这个异常对象，并处理；如果没有捕获 (catch) 这个异常对象，那么这个异常对象将会导致程序终止。

对于程序出现的异常，一般有两种解决方法：一是遇到借误就终止程序的运行。另一种方法是程序员在编写程序时，就充分考虑到各种可能发生的异常和错误，极力预防和避免。实在无法避免的，要编写相应的代码进行异常的检测。以及异常的处理，保证代码的健壮性。

`java.lang.Throwable`类是 Java 程序执行过程中发生的异常事件对应的类的根父类。

![image-20240310093431482](https://cdn.jsdelivr.net/gh/ZL85/ImageBed@main//202403100934658.png)

`Throwable`类可分为两类：Error 和 Exception，分别对应`java.lang.Eror`和`java.lang.Exception`两个类。

- [Error](https://docs.oracle.com/javase/8/docs/api/java/lang/Error.html)：Java 虚拟机无法解决的严重问题，如 JVM 系统内部错误、资源耗尽等严重情况，一般不编写针对性代码进行处理。

- [Exception](https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html)：其他因编程错误或偶然的外在因素导致的一般性问题，需要使用针对性的代码进行处理，使程序继续运行，如空指针异常、试图读取不存在的文件、网络连接中断、数组下标越界等。

开发中常见异常：

运行时异常：

- `ArrayIndexOutOfBoundsException`
- `NullPointerException`
- `ClassCastException`
- `NumberFormatException`
- `ClassCastException`
- `ArithmeticException`

编译时异常：

- `ClassNotFoundException`
- `FileNotFoundException`
- `IOException`

OOM异常。

```java
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.FileDescriptor;
import java.io.FileInputStream;
import java.util.Date;
import java.util.Scanner;

public class ExceptionTest {
    @Test
    public void test1() {
        int[] arr = new int[10];
        System.out.println(arr[10]); //java.lang.ArrayIndexOutOfBoundsException: Index 10 out of bounds for length 10
    }

    @Test
    public void test2() {
//        String str = "hello";
//        str = null;
//        System.out.println(str.toString()); //java.lang.NullPointerException: Cannot invoke "String.toString()" because "str" is null

//        int[] arr1 = new int[10];
//        arr1 = null;
//        System.out.println(arr1[0]); //java.lang.NullPointerException: Cannot load from int array because "arr" is null

        int[][] arr2 = new int[10][];
        System.out.println(arr2[0][0]); //java.lang.NullPointerException: Cannot load from int array because "arr2[0]" is null
    }

    @Test
    public void test3() {
        Object obj = new String();
        String str = (String) obj;

        Date date = (Date) obj; //java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.Date (java.lang.String and java.util.Date are in module java.base of loader 'bootstrap')
    }

    @Test
    public void test4() {
        String str = "123";
        str = "abc";
        int i = Integer.parseInt(str);
        System.out.println(i); //java.lang.NumberFormatException: For input string: "abc"
    }

    @Test
    public void test5() {
        Scanner scanner = new Scanner(System.in);
        int i = scanner.nextInt(); //InputMismatchException
        System.out.println(i);
    }

    @Test
    public void test6() {
        int i = 10;
        System.out.println(i / 0); //java.lang.ArithmeticException: / by zero
    }

    @Test
    public void test7() {
//        Class cls = Class.forName("java.lang.String"); //java.lang.ClassNotFoundException
    }

    @Test
    public void test8() {
//        File file = new File("hello.txt");
//
//        FileInputStream fis = new FileInputStream(file); //java.io.FileNotFoundException
//
//        int data = fis.read(); //java.io.IOException
//        while (data != -1) {
//            System.out.println((char) data);
//            data = fis.read(); //java.io.IOException
//        }
//        fis.close(); //java.io.IOException
    }
}
```

Java 采用的异常处理机制是将异常处理的程序集中在一起，与正常的程序代码分开，使得程序简洁，优雅，便于维护。

Java 异常处理方式：

- `try-catch-finally`
- `throws` + 异常类型
